### PR TITLE
Allow additional arguments to be passed into hooks

### PR DESCRIPTION
**Pull request by rhatdan**
If a packager wants to be able to support addititional arguments on his
hook this will allow them to setup the configuration with these arguments.

For example this would allow a hook developer to add support for a --debug
flag to change the level of debugging in his hook.

Signed-off-by: Daniel J Walsh <dwalsh@redhat.com>

**- What I did**

**- How I did it**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

**Comment by runcom** [comment]
Looks fine since OCI supports this right?

**Comment by rhatdan** [comment]
OCI Spec does not specify anything specific for args to the hook.  Just when to run it.

@mrunalp PTAL

**Comment by rhatdan** [comment]
Need to add a test for this to make sure it works.

**Comment by rhatdan** [comment]
Goal is to
Fix https://github.com/kubernetes-incubator/cri-o/issues/1222

**Comment by rhatdan** [comment]
@flx42 PTAL

**Comment by flx42** [comment]
As I mentioned in https://github.com/kubernetes-incubator/cri-o/issues/1222#issuecomment-354891207, I still have an issue with the fact that the stage is passed as `argv[1]`. This is not documented, right?

It seems a little weird to both pass the stage as `argv[1]` and allow the user to pass additional arguments.
Especially with this example, a flag like `--debug` should be before any subcommand. i.e. it should be `oci-umount --debug prestart` and not `oci-umount prestart --debug`.

I see multiple options now:
- Don't inject the stage as an argument. Instead, users must use this commit and manually add `prestart`, `poststart` or `poststop` to `arguments`. It will cost some duplication for hooks that support all cases.
- Don't inject the stage as an argument. Use an environment variable instead. This will remove duplication, but will require the users to modify their hook. Of course, this option is not mutually exclusive with the previous one, users will be able to choose which one is simpler.
- Inject the stage, but as the **last** argument on the command-line, after the user's specified arguments. Still not ideal though, but most command-line parsing tools will return you the first non-option argument, so a hook is less likely to choke on `./hook --debug --user --no-cgroups configure prestart`, than on `./hook prestart --debug --user --no-cgroups configure`

**Comment by rhatdan** [comment]
We could pass the Arguments before the Stage.

./hook $ARGUMENTS $STAGE

 ./hook --debug --user --no-cgroups configure prestart

Which would work, and not break existing hooks.

**Comment by rhatdan** [comment]
I don't see a way for the packager/admin to be able to easily handle your solution in the config, without them having to specify the arguments for each stage.

**Comment by rhatdan** [comment]
If I change the Arguments to Options, that might make it easier for others to understand.

**Comment by mrunalp** [comment]
What if we pass the stage as --stage stage_name?

**Comment by rhatdan** [comment]
@flx42 Would that be ok with you?  Hooks would have to handle or ignore the --stage $STAGE command which the container runtime would inject and then all other arguments would just be appended to the end?

**Comment by flx42** [comment]
> I don't see a way for the packager/admin to be able to  easily handle your solution in the config, without them having to  specify the arguments for each stage.

An environment variable is probably the only way.

> Would that be ok with you? Hooks would have to handle or ignore the --stage $STAGE command which the container runtime would inject and then all other arguments would just be appended to the end?

I think this is worse, since all the hooks need to be modified to handle this argument, even if it's just for skipping it. With `./hook --debug --user --no-cgroups configure prestart`, `flag.Args()` will be `[]string{"configure", "prestart"}`. No need to modify existing code, unless your program checks for trailing arguments.

**Comment by flx42** [comment]
To summarize, if we can't reach an agreement on a better solution. I would prefer that we either pass the stage as the last argument, or status quo.

**Comment by cyphar** [comment]
On 2018-01-04, Daniel J Walsh <reply@reply.github.com> wrote:
> OCI Spec does not specify anything specific for args to the hook.  Just when to run it.

That's not correct, the OCI configuration spec does specify how to pass
args to a hook[1], and it supported.

[1]: https://github.com/opencontainers/runtime-spec/blob/v1.0.1/config.md#posix-platform-hooks

--
Aleksa Sarai
Senior Software Engineer (Containers)
SUSE Linux GmbH
<https://www.cyphar.com/>

**Comment by flx42** [comment]
I think he simply meant that the OCI conf spec does not prohibit passing arguments to the hooks, but it doesn't specify any argument to pass either (like `prestart` as `argv[1]` assumed here).

**Comment by mrunalp** [comment]
@cyphar @flx42 Yep, there is no mechanism to pass the hook stage as in practice we have a single binary that implements different hook stages.

**Comment by mrunalp** [comment]
@flx42 We discussed passing the stage as an environment variable to the hook. What do you think of that?

**Comment by flx42** [comment]
Fine for me, but we also need this patch.

**Comment by mrunalp** [comment]
@flx42 Sounds good. Thanks!
 @rhatdan could you make the changes to pass stage as environment variable?

**Comment by rhatdan** [comment]
@mrunalp Yes, although it is going to be more involved then I hoped, since CRI-O uses opencontainers/runtime-tools and the released version of it does not support passing environment variables to the hooks.  
Hopefully will have something next week.

@cyphar We are talking about passing of $STAGE as Argv[1] in our hooks.  All of the hooks we (Red Hat) have written have assumed that Argv[1] would be the stage.  @flx42 has asked us to change that to make it an environment variable to set with the stage.  Not sure if this is something we should push to get into a newer version of the runtime spec, since the hooking knowing what stage it is being run in, is critical to being able to use the same hook for different stages.

**Comment by rhatdan** [comment]
Currently blocked by https://github.com/opencontainers/runtime-tools/pull/549

**Comment by rhatdan** [comment]
@flx42 @mrunalp @runcom PTAL

**Comment by mrunalp** [comment]
@rhatdan there are test failures in travis.

File: server/container_create.go [hunk 1]
```diff
@@ -976,21 +1039,39 @@ func (s *Server) createSandboxContainer(ctx context.Context, containerID string,
 			return nil, err
 		}
 
+		mnt = rspec.Mount{
+			Type:        "bind",
+			Source:      sb.ResolvPath(),
+			Destination: "/etc/resolv.conf",
+			Options:     options,
```
Comment (runcom) [inline server/container_create.go]: every call of `AddBindMount` that you've replaced with `AddMount`, you now need to add either `bind` (or `rbind`) to `options` when it's a bind mount (here for this case for instance https://github.com/kubernetes-incubator/cri-o/pull/1244/files#diff-dba9598e19f7903533dc9add98d410d2R1033). Cheers :)

Comment (runcom) [reply server/container_create.go]: You can take a look at how it was `AddBindMount` here https://github.com/opencontainers/runtime-tools/pull/279/files#diff-ed9e41f69ceb4f9d141e2fe7a9da68a4L914

Comment (runcom) [reply server/container_create.go]: @rhatdan ^^

**Comment by runcom** [comment]
> @rhatdan there are test failures in travis.

related to my comment above

**Comment by rhatdan** [comment]
@runcom Thanks, I added all of the "bind"s hopefully tests will pass.

**Comment by rhatdan** [comment]
Ok tests pass now,  @mrunalp @flx42 PTAL

**Comment by rhatdan** [comment]
@mrunalp @runcom PTAL

**Comment by rhatdan** [comment]
@mrunalp Seems test is hung, can I just restart this?

**Comment by rhatdan** [comment]
@mrunalp @runcom Passes all tests.  I will prepare the changes to the oci hooks.

**Comment by rhatdan** [comment]
@runcom PTAL

**Pull request merged**